library/summary.ml

```diff
@@ -140,26 +127,32 @@ let ref ?freeze ~name x = fst @@ ref_tag ?freeze ~name x
 
 module Local = struct
 
-type 'a local_ref = ('a CEphemeron.key * string) ref
+type 'a local_ref = ('a CEphemeron.key * 'a Dyn.tag) ref
 
-let (:=) r v = r := (CEphemeron.create v, snd !r)
+let set r v = r := (CEphemeron.create v, snd !r)
 
-let (!) r =
+let get r =
   let key, name = !r in
   try CEphemeron.get key
   with CEphemeron.InvalidKey ->
-    let { init_function } = String.Map.find name !sum_map in
+    let { init_function } = DynMap.find name !sum_map in
     init_function ();
     CEphemeron.get (fst !r)
 
 let ref ?(freeze=fun x -> x) ~name init =
-  let r = pervasives_ref (CEphemeron.create init, name) in
-  declare_summary name
-    { freeze_function = (fun ~marshallable -> freeze !r);
-      unfreeze_function = ((:=) r);
-      init_function = (fun () -> r := init) };
+  let () = check_name (mangle name) in
+  let tag : 'a Dyn.tag = Dyn.create (mangle name) in
+  let r = pervasives_ref (CEphemeron.create init, tag) in
+  let () = sum_map := DynMap.add tag
+    { freeze_function = (fun ~marshallable -> freeze (get r));
```

Comment:
This is broken!

You should not get the value on freeze, but rather the ephemeron key, as it was before.

Otherwise you defeat the purpose of this module, which is exactly to not freeze/marshal the value associated to the ephemeron key.

@ppedrot would you mind fixing this (I'm lost in the mess of these types, which, funnily enough, introduced an error rather than fixing one)

Reply:
I must be tired, but I don't see what this patch changed in the runtime. The code does the same thing as before using one less existential indirection. Before the closure was also calling `!` which has been renamed to get...

Reply:
Yes, apparently the problem predates this PR, sorry for the noise

Reply:
Well-typed program transformations can't be blamed!